File: backend/routes/tasks.py
from fastapi import APIRouter, HTTPException, Depends
from typing import Optional
from datetime import datetime, timezone
from database import db, get_next_key
from auth_helpers import get_current_user
from models import TaskInput, TaskUpdate
import uuid

router = APIRouter(prefix="/api/tasks", tags=["tasks"])

@router.get("")
async def get_tasks(project_id: Optional[str] = None, assignee_id: Optional[str] = None, status: Optional[str] = None, user=Depends(get_current_user)):
    query = {}
    if project_id:
        query["project_id"] = project_id
    if assignee_id:
        query["assignee_id"] = assignee_id
    if status:
        query["status"] = status
    return await db.tasks.find(query, {"_id": 0}).sort("created_at", -1).to_list(500)

@router.get("/my")
async def get_my_tasks(user=Depends(get_current_user)):
    tasks = await db.tasks.find({"assignee_id": user["_id"]}, {"_id": 0}).sort("created_at", -1).to_list(200)
    for t in tasks:
        proj = await db.projects.find_one({"id": t["project_id"]}, {"_id": 0, "name": 1, "key": 1, "color": 1})
        t["project"] = proj
    return tasks

@router.post("")
async def create_task(input: TaskInput, project_id: str, user=Depends(get_current_user)):
    project = await db.projects.find_one({"id": project_id}, {"_id": 0})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    key = await get_next_key(project_id, project["key"])
    doc = {
        "id": str(uuid.uuid4()), "project_id": project_id, "key": key,
        "title": input.title, "description": input.description,
        "status": input.status, "priority": input.priority,
        "assignee_id": input.assignee_id, "due_date": input.due_date,
        "created_by": user["_id"],
        "created_by_name": user.get("name", ""),
        "created_at": datetime.now(timezone.utc).isoformat(),
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "comment_count": 0,
    }
    await db.tasks.insert_one(doc)
    doc.pop("_id", None)
    # Log "created" activity
    await db.comments.insert_one({
        "id": str(uuid.uuid4()), "entity_id": doc["id"], "entity_type": "task",
        "user_id": user["_id"], "user_name": user.get("name", ""),
        "content": "created this task", "type": "activity",
        "created_at": datetime.now(timezone.utc).isoformat(),
    })
    if input.assignee_id and input.assignee_id != user["_id"]:
        await db.notifications.insert_one({
            "id": str(uuid.uuid4()), "user_id": input.assignee_id,
            "type": "task_assigned", "message": f"You were assigned to {key}: {input.title}",
            "link": f"/project/{project_id}/board", "read": False,
            "created_at": datetime.now(timezone.utc).isoformat(),
        })
    return doc

@router.put("/{task_id}")
async def update_task(task_id: str, input: TaskUpdate, user=Depends(get_current_user)):
    old = await db.tasks.find_one({"id": task_id}, {"_id": 0})
    if not old:
        raise HTTPException(status_code=404, detail="Task not found")
    update_data = {k: v for k, v in input.model_dump().items() if v is not None}
    update_data["updated_at"] = datetime.now(timezone.utc).isoformat()
    await db.tasks.update_one({"id": task_id}, {"$set": update_data})
    # Log status change activity
    if "status" in update_data and update_data["status"] != old.get("status"):
        await db.comments.insert_one({
            "id": str(uuid.uuid4()), "entity_id": task_id, "entity_type": "task",
            "user_id": user["_id"], "user_name": user.get("name", ""),
            "content": f"changed status to {update_data['status'].replace('_', ' ').title()}",
            "type": "activity", "created_at": datetime.now(timezone.utc).isoformat(),
        })
        target = old.get("assignee_id") or old.get("created_by")
        if target and target != user["_id"]:
            await db.notifications.insert_one({
                "id": str(uuid.uuid4()), "user_id": target,
                "type": "status_changed", "message": f"{old['key']} status changed to {update_data['status']}",
                "link": f"/project/{old['project_id']}/board", "read": False,
                "created_at": datetime.now(timezone.utc).isoformat(),
            })
    # Log assignee change activity
    if "assignee_id" in update_data and update_data["assignee_id"] != old.get("assignee_id"):
        if update_data["assignee_id"]:
            assignee_user = await db.users.find_one({"_id": __import__('bson').ObjectId(update_data["assignee_id"])})
            assignee_name = assignee_user.get("name", "someone") if assignee_user else "someone"
            await db.comments.insert_one({
                "id": str(uuid.uuid4()), "entity_id": task_id, "entity_type": "task",
                "user_id": user["_id"], "user_name": user.get("name", ""),
                "content": f"assigned this to {assignee_name}",
                "type": "activity", "created_at": datetime.now(timezone.utc).isoformat(),
            })
            if update_data["assignee_id"] != user["_id"]:
                await db.notifications.insert_one({
                    "id": str(uuid.uuid4()), "user_id": update_data["assignee_id"],
                    "type": "task_assigned", "message": f"You were assigned to {old['key']}: {old['title']}",
                    "link": f"/project/{old['project_id']}/board", "read": False,
                    "created_at": datetime.now(timezone.utc).isoformat(),
                })
    return await db.tasks.find_one({"id": task_id}, {"_id": 0})

@router.delete("/{task_id}")
async def delete_task(task_id: str, user=Depends(get_current_user)):
    await db.tasks.delete_one({"id": task_id})
    await db.comments.delete_many({"entity_id": task_id})
    return {"message": "Task deleted"}

File: backend/routes/tasks.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const Task = require('../models/Task');
const Project = require('../models/Project');
const Comment = require('../models/Comment');
const Notification = require('../models/Notification');
const Counter = require('../models/Counter');
const User = require('../models/User');
const auth = require('../middleware/auth');

const router = express.Router();

// GET /api/tasks
router.get('/', auth, async (req, res) => {
  try {
    const query = {};
    if (req.query.project_id) query.project_id = req.query.project_id;
    if (req.query.assignee_id) query.assignee_id = req.query.assignee_id;
    if (req.query.status) query.status = req.query.status;
    const tasks = await Task.find(query).select('-_id -__v').sort({ created_at: -1 }).limit(500).lean();
    res.json(tasks);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// GET /api/tasks/my
router.get('/my', auth, async (req, res) => {
  try {
    const tasks = await Task.find({ assignee_id: req.user._id }).select('-_id -__v').sort({ created_at: -1 }).limit(200).lean();
    for (const t of tasks) {
      const proj = await Project.findOne({ id: t.project_id }).select('name key color -_id').lean();
      t.project = proj;
    }
    res.json(tasks);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// POST /api/tasks?project_id=X
router.post('/', auth, async (req, res) => {
  try {
    const { project_id } = req.query;
    const project = await Project.findOne({ id: project_id }).lean();
    if (!project) return res.status(404).json({ detail: 'Project not found' });

    const key = await Counter.getNextKey(project_id, project.key);
    const { title, description = '', status = 'backlog', priority = 'medium', assignee_id = null, due_date = null } = req.body;
    const now = new Date().toISOString();
    const taskId = uuidv4();

    const doc = await Task.create({
      id: taskId, project_id, key, title, description, status, priority,
      assignee_id, due_date, created_by: req.user._id, created_by_name: req.user.name,
      created_at: now, updated_at: now,
    });

    // Activity log
    await Comment.create({
      id: uuidv4(), entity_id: taskId, entity_type: 'task',
      user_id: req.user._id, user_name: req.user.name,
      content: 'created this task', type: 'activity', created_at: now,
    });

    // Notification
    if (assignee_id && assignee_id !== req.user._id) {
      await Notification.create({
        id: uuidv4(), user_id: assignee_id, type: 'task_assigned',
        message: `You were assigned to ${key}: ${title}`,
        link: `/project/${project_id}/board`, created_at: now,
      });
    }

    const result = doc.toObject();
    delete result._id; delete result.__v;
    res.json(result);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// PUT /api/tasks/:id
router.put('/:id', auth, async (req, res) => {
  try {
    const old = await Task.findOne({ id: req.params.id }).select('-_id -__v').lean();
    if (!old) return res.status(404).json({ detail: 'Task not found' });

    const updateData = {};
    for (const [k, v] of Object.entries(req.body)) {
      if (v !== undefined && v !== null) updateData[k] = v;
    }
    updateData.updated_at = new Date().toISOString();

    await Task.updateOne({ id: req.params.id }, { $set: updateData });
    const now = new Date().toISOString();

    // Status change activity
    if (updateData.status && updateData.status !== old.status) {
      await Comment.create({
        id: uuidv4(), entity_id: req.params.id, entity_type: 'task',
        user_id: req.user._id, user_name: req.user.name,
        content: `changed status to ${updateData.status.replace(/_/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}`,
        type: 'activity', created_at: now,
      });
      const target = old.assignee_id || old.created_by;
      if (target && target !== req.user._id) {
        await Notification.create({
          id: uuidv4(), user_id: target, type: 'status_changed',
          message: `${old.key} status changed to ${updateData.status}`,
          link: `/project/${old.project_id}/board`, created_at: now,
        });
      }
    }

    // Assignee change activity
    if (updateData.assignee_id && updateData.assignee_id !== old.assignee_id) {
      const assigneeUser = await User.findById(updateData.assignee_id);
      const assigneeName = assigneeUser?.name || 'someone';
      await Comment.create({
        id: uuidv4(), entity_id: req.params.id, entity_type: 'task',
        user_id: req.user._id, user_name: req.user.name,
        content: `assigned this to ${assigneeName}`, type: 'activity', created_at: now,
      });
      if (updateData.assignee_id !== req.user._id) {
        await Notification.create({
          id: uuidv4(), user_id: updateData.assignee_id, type: 'task_assigned',
          message: `You were assigned to ${old.key}: ${old.title}`,
          link: `/project/${old.project_id}/board`, created_at: now,
        });
      }
    }

    const updated = await Task.findOne({ id: req.params.id }).select('-_id -__v').lean();
    res.json(updated);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// DELETE /api/tasks/:id
router.delete('/:id', auth, async (req, res) => {
  try {
    await Task.deleteOne({ id: req.params.id });
    await Comment.deleteMany({ entity_id: req.params.id });
    res.json({ message: 'Task deleted' });
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

module.exports = router;
